coffee/rewriter.py

```diff
@@ -131,132 +145,131 @@ def licm(self, merge_and_simplify=False, compact_tmps=False):
-        self.stmt.children[1] = Par(ast_c_sum(new_prods))
+                     * mode == 'standard': this simple heuristics consists of \
+                                           grouping on symbols that appear the \
+                                           most in the expression.
+                     * mode == 'immutable': if many static constant objects are \
+                                            expected, with this strategy they are \
+                                            grouped together, within the obvious \
+                                            limits imposed by the expression itself.
+        """
+        info = visit(self.stmt.children[1])
+        symbols_dep = info['symbols_dep']
+
+        # Select the factorization strategy
+        if mode == 'standard':
+            occurrences = [s.rank for s in symbols_dep.keys()]
+            occurrences = dict((i, occurrences.count(i)) for i in occurrences if i)
+            dimension = max(occurrences.iteritems(), key=operator.itemgetter(1))[0]
+            should_factorize = lambda n: n.rank == dimension
+        elif mode == 'immutable':
+            should_factorize = lambda n: self.decls[n.symbol].is_static_const
+        if mode not in ['standard', 'immutable']:
+            warning('Unknown factorization strategy. Skipping.')
+            return
 
+        # Perform the factorization
+        self.expr_factorizer.factorize(should_factorize)
 
-class ExpressionHoister(object):
-    """Perform loop-invariant code motion (licm).
+    @staticmethod
+    def reset():
+        ExpressionHoister._expr_handled[:] = []
+        ExpressionExpander._expr_handled[:] = []
 
-    Invariant expressions found in the loop nest are moved "after" the
-    outermost independent loop and "after" the fastest varying dimension
-    loop. Here, "after" means that if the loop nest has two loops ``i``
-    and ``j``, and ``j`` is in the body of ``i``, then ``i`` comes after
-    ``j`` (i.e. the loop nest has to be read from right to left).
 
-    For example, if a sub-expression ``E`` depends on ``[i, j]`` and the
-    loop nest has three loops ``[i, j, k]``, then ``E`` is hoisted out from
-    the body of ``k`` to the body of ``i``). All hoisted expressions are
-    then wrapped within a suitable loop in order to exploit compiler
-    autovectorization. Note that this applies to constant sub-expressions
-    as well, in which case hoisting after the outermost loop takes place."""
+class ExpressionHoister(object):
 
     # Track all expressions to which LICM has been applied
     _expr_handled = []
     # Temporary variables template
     _hoisted_sym = "%(loop_dep)s_%(expr_id)d_%(round)d_%(i)d"
 
+    # Constants used by the extract method to charaterize expressions:
+    INV = 0  # /INVariant/: expression is loop invariant
+    KSE = 1  # /KeepSEarching/: expression is potentially in a larger invariant
+    HOI = 2  # /HOIsted/: expression is *not* invariant, should be hoisted
```

Comment:
Do you mean "should not be hoisted any further"?

I know you have an inordinate love of short variable names, but I always feels like descriptive named constants ought to be so.